File: cpp/sieve.cpp
// Sieve implementation.

#include "sieve.hpp"

namespace euler {

void
Sieve::fill(int nsize) {
  primes.assign(nsize, true);

  primes[0] = false;
  primes[1] = false;
  int pos = 2;
  while (pos < nsize) {
    if (primes[pos]) {
      int n = pos + pos;
      while (n < nsize) {
	primes[n] = false;
	n += pos;
      }
      if (pos == 2)
	pos = 3;
      else
	pos += 2;
    } else {
      pos += 2;
    }
  }
}

bool
Sieve::is_prime(int num) {
  if (num >= (int)primes.size())
    new_length(num);

  return primes[num];
}

void
Sieve::new_length(int n) {
  int cur = primes.size();
  while (n >= cur)
    cur *= 8;
  fill(cur);
}

int
Sieve::next_prime(int num) {
  if (num == 2)
    return 3;
  do
    num += 2;
  while (!is_prime(num));

  return num;
}

}

File: swift/Sources/sieve.swift
// A prime sieve.

public class Sieve {
    var vec: Array<Bool>
    var limit: Int

    init(isize: Int = 8192) {
        limit = isize
        vec = Array(repeating: true, count: limit + 1)
        fill()
    }

    public func isPrime(_ n: Int) -> Bool {
        if n > limit {
            var newLimit = limit
            while newLimit < n {
                newLimit *= 8
            }

            vec = Array(repeating: true, count: newLimit + 1)
            limit = newLimit
            fill()
        }

        return vec[n]
    }

    public func nextPrime(_ n: Int) -> Int {
        if n == 2 {
            return 3;
        }

        var next = n + 2
        while !isPrime(next) {
            next += 2
        }
        return next
    }

    // TODO: Can we make this work as a sequence?
    public func generate(_ upto: Int) -> SieveIterator {
        return SieveIterator(sieve: self, upto: upto)
    }

    func fill() {
        vec[0] = false
        vec[1] = false

        var pos = 2
        while pos <= limit {
            if !vec[pos] {
                pos += 2
            } else {
                var n = pos + pos
                while n <= limit {
                    vec[n] = false
                    n += pos
                }
                if pos == 2 {
                    pos += 1
                } else {
                    pos += 2
                }
            }
        }
    }
}

public class SieveIterator: Sequence, IteratorProtocol {
    public typealias Element = Int
    var sieve: Sieve
    var cur: Int
    var limit: Int

    init(sieve: Sieve, upto: Int) {
        self.sieve = sieve
        cur = 2
        self.limit = 2
    }

    public func next() -> Int? {
        if cur > limit {
            return nil
        }

        let result = cur
        cur = sieve.nextPrime(cur)
        return result
    }
}
